packages/opentelemetry-tracing/src/export/BatchSpanProcessor.ts

```diff
@@ -82,22 +74,33 @@ export class BatchSpanProcessor implements SpanProcessor {
   /** Add a span in the buffer. */
   private _addToBuffer(span: ReadableSpan) {
     this._finishedSpans.push(span);
+    this._maybeStartTimer();
     if (this._finishedSpans.length > this._bufferSize) {
       this._flush();
     }
   }
 
-  private _shouldFlush(): boolean {
-    return (
-      this._finishedSpans.length >= 0 &&
-      Date.now() - this._lastSpanFlush >= this._bufferTimeout
-    );
-  }
-
   /** Send the span data list to exporter */
   private _flush() {
+    this._clearTimer();
+    if (this._finishedSpans.length === 0) return;
     this._exporter.export(this._finishedSpans, () => {});
     this._finishedSpans = [];
-    this._lastSpanFlush = Date.now();
+  }
+
+  private _maybeStartTimer() {
```

Comment:
Personally, I started preferring the early `return`. Something like this:

```js
  private _maybeStartTimer() {
    if (this._timer !== undefined) return;

    this._timer = setTimeout(() => {
      this._flush();
    }, this._bufferTimeout);
    unrefTimer(this._timer);
  }
```
Feel free to ignore if you disagree.